### PR TITLE
Feature/self service by profile

### PR DESCRIPTION
**Pull request by whitx**
# Description
Allow the creation of device registration profile to be associated with connection profile

# Impacts
Device registration workflow, connection profile workflow

# Delete branch after merge
NO

# NEWS file entries
## Enhancements
* Improved device registration to be able to create profiles, this allow you to associate device registration profile by connection profile

# UPGRADE file entries
You will need to remove anything related [device_registration] in the conf/pf.conf file.

### CODE REVIEW

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 1]
```diff
@@ -1851,21 +1851,15 @@ PacketFence allows users to extend their access before it has ended. In order to
...
+As from PacketFence 7.3 and onwards you need to create a device registration profile and associate it with the connection profile.
 
-   [registration]
-   device_registration = enabled
-   device_registration_role = gaming
-
-Make sure the role exists in PacketFence otherwise you will encounter registration errors. Moreover, make sure the role mapping for your particular equipment is done.
-
-These parameters can also be configured from the *Configuration -> -> Advanced Access Configuration -> Device Registration* section.
+This will allow you to configure who is able to register which type of device. For instance a person connection from the public WiFi can only register Android but a person from the intern network can register Windows and Mac.
```
Comment (extrafu) [inline docs/PacketFence_Administration_Guide.asciidoc]: For instance a person connection from the public WiFi can only register Android but a person from the intern network can register Windows and Mac. => For instance a person connecting from the public WiFi can only register an Android device but a person from the internal network can register Windows and MacOS computers.

File: html/pfappserver/lib/pfappserver/Form/Config/DeviceRegistration.pm [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+has roles => ( is => 'rw' );
+
+## Definition
+has_field 'id' =>
+  (
+   type => 'Text',
+   label => 'Device Registration ID',
+   required => 1,
+   messages => { required => 'Please specify the ID of the Device Registration entry.' },
```
Comment (extrafu) [inline html/pfappserver/lib/pfappserver/Form/Config/DeviceRegistration.pm]: Not very clear

File: html/pfappserver/lib/pfappserver/Form/Config/DeviceRegistration.pm [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+has_field 'allowed_devices' =>
+  (
+   type => 'FingerbankSelect',
+   multiple => 1,
+   label => 'OS',
+   element_class => ['chzn-deselect'],
+   element_attr => {'data-placeholder' => 'Click to add an OS'},
+   tags => { after_element => \&help,
+             help => 'List of OS which will be allowed to register via the self service portal.' },
```
Comment (extrafu) [inline html/pfappserver/lib/pfappserver/Form/Config/DeviceRegistration.pm]: will be allowed to be registered

**Comment by fdurand** [comment]
I like this feature, just need to test it before merging

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Status.pm [hunk 4]
```diff
@@ -54,7 +54,7 @@ sub index : Path : Args(0) {
     if (view($pid)) {
         $c->stash->{hasLocalAccount} = $TRUE;
     }
-    if (isenabled( $Config{'device_registration'}{'status'} ) ) {
+    if (defined( $c->profile{'device_registration'} ) ) {
```
Comment (fdurand) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Status.pm]: $c->profile->{'device_registration'}   ?

File: html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm [hunk 5]
```diff
@@ -392,6 +393,33 @@ has_field 'scans.contains' =>
     widget_wrapper => 'DynamicTableRow',
   );
 
+=head2 device_registration
+
+Collection Device registration profiles for the profile
+
+=cut
+
+has_field 'devices_registration' =>
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm]: device_registration or devices_registration ?

File: html/pfappserver/lib/pfappserver/PacketFence/Controller/Config/DeviceRegistration.pm [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+=head1 DESCRIPTION
+
+ConnectionDeviceRegistration
+
+=cut
+
+use HTTP::Status qw(:constants is_error is_success);
+use Moose;  # automatically turns on strict and warnings
+use namespace::autoclean;
+#use pf::factory::device_registration;
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/PacketFence/Controller/Config/DeviceRegistration.pm]: remove

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Status.pm (hunk 4 above)
Comment (fdurand) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Status.pm]: $c->profile->{'_devices_registration'}

**Comment by fdurand** [comment]
Still have $Config{'device_registration'}{'status'} in the code

**Comment by fdurand** [comment]
Tested and works.
It just miss the packaging stuff.

File: addons/packages/packetfence.spec [hunk 7]
```diff
@@ -877,6 +877,9 @@ fi
 %config                 /usr/local/pf/conf/chi.conf.defaults
 %config(noreplace)      /usr/local/pf/conf/portal_modules.conf
 %config                 /usr/local/pf/conf/portal_modules.conf.defaults
+%config(noreplace)      /usr/local/pf/conf/device_registration.conf
+%config                 /usr/local/pf/conf/device_registration.conf.defaults
+%config                 /usr/local/pf/conf/device_registration.conf.example
```
Comment (fdurand) [inline addons/packages/packetfence.spec]: Remove %config for example file

File: html/pfappserver/lib/pfappserver/Form/Config/DeviceRegistration.pm [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+has roles => ( is => 'rw' );
+
+## Definition
+has_field 'id' =>
+  (
+   type => 'Text',
+   label => 'Profile Name',
+   required => 1,
+   messages => { required => 'Please specify an name of the Device Registration entry.' },
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Form/Config/DeviceRegistration.pm]: Please specify a name ...

File: html/pfappserver/lib/pfappserver/Form/Config/DeviceRegistration.pm [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+   apply => [ pfappserver::Base::Form::id_validator('device registration ID') ]
+  );
+
+has_field 'description' =>
+  (
+   type => 'Text',
+   messages => { required => 'Please specify the Description of the Device Registration entry.' },
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Form/Config/DeviceRegistration.pm]: Please specify the description.

File: html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm [hunk 10]
```diff
@@ -392,6 +393,33 @@ has_field 'scans.contains' =>
     widget_wrapper => 'DynamicTableRow',
   );
 
+=head2 device_registration
+
+Collection Device registration profiles for the profile
+
+=cut
+
+has_field 'device_registration' =>
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Form/Config/ProfileCommon.pm]: Can it be unique ?

File: html/pfappserver/lib/pfappserver/Model/Config/DeviceRegistration.pm [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
+USA.
+
+=cut
+
+1;
+
+
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Model/Config/DeviceRegistration.pm]: Remove spaces

**Review by fdurand — CHANGES_REQUESTED** [review]

**Comment by fdurand** [comment]
ERROR: [mac:unknown] Caught exception in pfappserver::Controller::Config::Profile->view "isa check for "cachedConfig" failed: Use of uninitialized value in subroutine entry at /usr/local/pf/lib/pf/ConfigStore.pm line 34." (pfappserver::PacketFence::Controller::Root::end)

**Pull request merged**